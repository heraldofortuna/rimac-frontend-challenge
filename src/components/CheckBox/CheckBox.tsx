import type { ICheckBoxProps } from '../../types/components/checkbox';
import styles from './CheckBox.module.scss';

const CheckBox: React.FC<ICheckBoxProps> = ({ 
  label,
  checked,
  onChange
}) => {
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    onChange(event.target.checked);
  };

  return (
    <label className={styles.checkbox}>
      <span>{label}</span>
      <input 
        type="checkbox" 
        checked={checked}
        onChange={handleChange}
        className={styles.checkbox__input} 
      />
      <span className={styles.checkbox__checkmark}></span>
    </label>
  );
};

export default CheckBox;