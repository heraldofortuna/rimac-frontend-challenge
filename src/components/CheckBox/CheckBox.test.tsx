import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import CheckBox from "./CheckBox";
import type { ICheckBoxProps } from "../../types/components/checkbox";

describe("CheckBox component", () => {
  const defaultProps: ICheckBoxProps = {
    label: "Test Checkbox",
    checked: false,
    onChange: vi.fn(),
  };

  const renderCheckBox = (props: Partial<ICheckBoxProps> = {}) => {
    return render(<CheckBox {...defaultProps} {...props} />);
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders correctly with default props", () => {
    renderCheckBox();
    
    const checkbox = screen.getByLabelText("Test Checkbox");
    expect(checkbox).toBeInTheDocument();
    expect(checkbox).toHaveAttribute("type", "checkbox");
    expect(checkbox).not.toBeChecked();
    expect(screen.getByText("Test Checkbox")).toBeInTheDocument();
  });

  it("renders as checked when checked prop is true", () => {
    renderCheckBox({ checked: true });
    
    const checkbox = screen.getByLabelText("Test Checkbox");
    expect(checkbox).toBeChecked();
  });

  it("calls onChange handler with correct value when clicked", async () => {
    const handleChange = vi.fn();
    renderCheckBox({ onChange: handleChange });
    
    const checkbox = screen.getByLabelText("Test Checkbox");
    fireEvent.click(checkbox);
    
    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(handleChange).toHaveBeenCalledWith(true);
  });

  it("toggles checked state when clicked", async () => {
    const handleChange = vi.fn();
    const { rerender } = renderCheckBox({ onChange: handleChange });
    
    const checkbox = screen.getByLabelText("Test Checkbox");
    
    fireEvent.click(checkbox);
    expect(handleChange).toHaveBeenCalledWith(true);

    rerender(<CheckBox {...defaultProps} checked={true} onChange={handleChange} />);

    fireEvent.click(checkbox);
    expect(handleChange).toHaveBeenCalledWith(false);
  });

  it("renders correctly with custom label", () => {
    renderCheckBox({ label: "Custom Label" });
    
    expect(screen.getByText("Custom Label")).toBeInTheDocument();
    expect(screen.getByLabelText("Custom Label")).toBeInTheDocument();
  });

  it("matches snapshot when unchecked", () => {
    const { asFragment } = renderCheckBox();
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot when checked", () => {
    const { asFragment } = renderCheckBox({ checked: true });
    expect(asFragment()).toMatchSnapshot();
  });
});