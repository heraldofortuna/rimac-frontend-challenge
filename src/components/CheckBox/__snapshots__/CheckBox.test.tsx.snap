// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CheckBox component > matches snapshot when checked 1`] = `
<DocumentFragment>
  <label
    class="_checkbox_b4c44f"
  >
    <span>
      Test Checkbox
    </span>
    <input
      checked=""
      class="_checkbox__input_b4c44f"
      type="checkbox"
    />
    <span
      class="_checkbox__checkmark_b4c44f"
    />
  </label>
</DocumentFragment>
`;

exports[`CheckBox component > matches snapshot when unchecked 1`] = `
<DocumentFragment>
  <label
    class="_checkbox_b4c44f"
  >
    <span>
      Test Checkbox
    </span>
    <input
      class="_checkbox__input_b4c44f"
      type="checkbox"
    />
    <span
      class="_checkbox__checkmark_b4c44f"
    />
  </label>
</DocumentFragment>
`;
